core/src/main/java/org/apache/iceberg/SerializedTable.java

```diff
@@ -0,0 +1,270 @@
+    this.defaultSortOrderId = table.sortOrder().orderId();
+    this.sortOrders = copy(table.sortOrders());
+    this.io = io;
+    this.encryption = table.encryption();
+    this.locationProvider = table.locationProvider();
+  }
+
+  private Table lazyTable() {
+    if (lazyTable == null) {
+      synchronized (this) {
+        if (lazyTable == null) {
+          TableOperations ops = new StaticTableOperations(metadataFileLocation, io, locationProvider);
+          this.lazyTable = newTable(ops, name);
+        }
+      }
+    }
+
+    return lazyTable;
+  }
+
+  protected Table newTable(TableOperations ops, String tableName) {
+    return new BaseTable(ops, tableName);
+  }
+
+  private String metadataFileLocation(Table table) {
+    if (table instanceof HasTableOperations) {
+      TableOperations ops = ((HasTableOperations) table).operations();
+      return ops.current().metadataFileLocation();
+
+    } else if (table instanceof BaseMetadataTable) {
+      return ((BaseMetadataTable) table).metadataLocation();
+
+    } else {
+      throw new IllegalArgumentException("Cannot determine metadata file location for table " + table.name());
+    }
+  }
+
+  private <K, V> Map<K, V> copy(Map<K, V> map) {
+    Map<K, V> copy = Maps.newHashMap(map);
+    return Collections.unmodifiableMap(copy);
+  }
+
+  @Override
+  public String name() {
+    return name;
+  }
+
+  @Override
+  public String location() {
+    return location;
+  }
+
+  @Override
+  public Map<String, String> properties() {
+    return properties;
+  }
+
+  @Override
+  public Schema schema() {
+    return schema;
+  }
+
+  @Override
+  public PartitionSpec spec() {
+    return specs.get(defaultSpecId);
+  }
+
+  @Override
+  public Map<Integer, PartitionSpec> specs() {
+    return specs;
+  }
+
+  @Override
+  public SortOrder sortOrder() {
+    return sortOrders.get(defaultSortOrderId);
+  }
+
+  @Override
+  public Map<Integer, SortOrder> sortOrders() {
+    return sortOrders;
+  }
+
+  @Override
+  public FileIO io() {
+    return io;
+  }
+
+  @Override
+  public EncryptionManager encryption() {
+    return encryption;
+  }
+
+  @Override
+  public LocationProvider locationProvider() {
+    return locationProvider;
+  }
+
+  @Override
+  public void refresh() {
+    throw new UnsupportedOperationException(errorMsg("refresh"));
+  }
+
+  @Override
+  public TableScan newScan() {
+    return lazyTable().newScan();
+  }
+
+  @Override
+  public Snapshot currentSnapshot() {
+    return lazyTable().currentSnapshot();
+  }
+
+  @Override
+  public Snapshot snapshot(long snapshotId) {
+    return lazyTable().snapshot(snapshotId);
+  }
+
+  @Override
+  public Iterable<Snapshot> snapshots() {
+    return lazyTable().snapshots();
+  }
+
+  @Override
+  public List<HistoryEntry> history() {
+    return lazyTable().history();
+  }
+
+  @Override
+  public UpdateSchema updateSchema() {
+    throw new UnsupportedOperationException(errorMsg("updateSchema"));
```

Comment:
Should we fall back to letting the `StaticTableOperations` throw the exception for this? I could go either way. If we do it here then it is clear that the problem is that the table was serialized, not just that the ops is static.

Reply:
I am ok either way. I think a slightly more descriptive error message is worth it in this case.

Reply:
That was addressed in a separate PR. No longer applies here.